File: codes/03_func/random_dist/with_skewness/02_rng_with_skewness.py
import numpy as np
import pylab as pl
from scipy.stats import skew


def rand_skew_norm(alpha, location, scale):
    sigma = alpha / np.sqrt(1.0 + alpha**2)

    u0, v0 = np.random.randn(2)
    u1 = sigma*u0 + np.sqrt(1.0 - sigma**2) * v0

    if u0 >= 0:
        return u1*scale + location
    return (-u1)*scale + location
# ------------------------------------------------------------------#


def randn_skew(N, alpha=0.0, location=0.0, scale=1.0):
    return [rand_skew_norm(alpha, location, scale) for x in range(N)]
# ------------------------------------------------------------------#


def randn_skew_fast(N, alpha=0.0, loc=0.0, scale=1.0):
    sigma = alpha / np.sqrt(1.0 + alpha**2)
    u0 = np.random.randn(N)
    v = np.random.randn(N)
    u1 = (sigma*u0 + np.sqrt(1.0 - sigma**2)*v) * scale
    u1[u0 < 0] *= -1
    u1 = u1 + loc
    return u1
# ------------------------------------------------------------------#


def skweness_to_alpha(gamma):

    assert(np.abs(gamma) < 0.995217)

    a = np.abs(gamma)**(2.0/3.0)
    denominator = 1.0/(a + (0.5*(4.0 - np.pi))**(2.0/3.0))
    delta = np.sqrt(0.5*np.pi*a*denominator)

    alpha = delta/np.sqrt(1.0-delta*delta)
    if gamma >= 0:
        return alpha
    return -alpha


NUM_SAMPLES = 100000
# skweness is between -0.9952717, -0.9952717
SKEWNESS = [-0.5, 0, 0.5]
ALPHA = [skweness_to_alpha(i) for i in SKEWNESS]

# lets check they at least visually match the PDF:
pl.subplots(figsize=(12, 4))
for i in range(len(ALPHA)):
    p = randn_skew(NUM_SAMPLES, ALPHA[i])
    # p = randn_skew_fast(NUM_SAMPLES, ALPHA[i])
    print "skewness = %g" % skew(p)
    pl.hist(p, bins=100, alpha=0.5, label=str(SKEWNESS[i]), density=True)

pl.title("histogram")
pl.ylabel("probability")
pl.xlabel("value")
pl.legend()
pl.savefig("fig1.png", dpi=150)
# pl.show()

File: codes/03_func/random_dist/with_skewness/02_rng_with_skewness.cpp
#include <iostream>
#include <stdio.h>
#include <vector>
#include <cmath>
#include <random>
#include <assert.h>

#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>
#include <gsl/gsl_sf_gamma.h>

#define RANDOM_GAUSS(S) gsl_ran_gaussian(gsl_rng_r, S)
#define INITIALIZE_RANDOM_CLOCK(seed)              \
    {                                              \
        gsl_rng_env_setup();                       \
        if (!getenv("GSL_RNG_SEED"))               \
            gsl_rng_default_seed = time(0) + seed; \
        gsl_rng_T = gsl_rng_default;               \
        gsl_rng_r = gsl_rng_alloc(gsl_rng_T);      \
    }
#define INITIALIZE_RANDOM_F(seed)             \
    {                                         \
        gsl_rng_env_setup();                  \
        if (!getenv("GSL_RNG_SEED"))          \
            gsl_rng_default_seed = seed;      \
        gsl_rng_T = gsl_rng_default;          \
        gsl_rng_r = gsl_rng_alloc(gsl_rng_T); \
    }
#define FREE_RANDOM gsl_rng_free(gsl_rng_r)

static const gsl_rng_type *gsl_rng_T;
static gsl_rng *gsl_rng_r;

using std::cout;
using std::endl;
using std::vector;

double rng_normal_skewed(const double alpha,
                         const double loc,
                         const double scale)
{
    // alpha control the skewness but it is different
    // loc and scale are the same parameters as in normal distribution

    double sigma = alpha / sqrt(1.0 + alpha * alpha);

    double u0 = RANDOM_GAUSS(1.0);
    double v0 = RANDOM_GAUSS(1.0);
    double u1 = sigma * u0 + sqrt(1.0 - sigma * sigma) * v0;

    if (u0 >= 0)
        return (u1 * scale + loc);
    return (-u1 * scale + loc);
}



// to calculate parameter alpha from skewness
double skewness_to_alpha(const double gamma)
{
    // The maximum (theoretical) skewness is obtained by setting delta =1 
    // in the skewness equation, giving gamma 0.9952717
    
    assert(std::abs(gamma) < 0.995217);

    double a = pow(std::abs(gamma), 2.0 / 3.0);
    double b = 0.5 * (4.0 - M_PI);
    double denom = 1.0 / (a + pow(b, 2.0 / 3.0));
    double delta = sqrt(0.5 * M_PI * a * denom);

    double alpha = delta / sqrt(1.0 - delta * delta);
    if (gamma >= 0)
        return alpha;
    return (-alpha);
}

int main()
{

    size_t n = 10000;
    constexpr double location = 0.0;
    constexpr double scale = 1.0;
    constexpr double skewness = 0.9;
    const double alpha = skewness_to_alpha(skewness);

    const int seed = 124;

    INITIALIZE_RANDOM_F(seed);
    for (size_t i = 0; i < n; i++)
    {
        cout << rng_normal_skewed(alpha, location, scale) << endl;
    }
    FREE_RANDOM;
}



// g++ 02_rng_with_skewness.cpp -lgsl -lgslcblas
// ./a.out>t
// plot in python
// pl.hist(np.loadtxt("t"), bins=100,  density=True); pl.show()


// to claculate the skewness
// from scipy.stats import skew
// skew(random_number_array)
